allennlp/models/language_model.py

```diff
@@ -0,0 +1,206 @@
+    relu: bool, optional (default=``True``)
+        whether add a ReLU non-linear transformation before the softmax layer
+    initializer : ``InitializerApplicator``, optional (default=``InitializerApplicator()``)
+        Used to initialize the model parameters.
+    regularizer : ``RegularizerApplicator``, optional (default=``None``)
+        If provided, will be used to calculate the regularization penalty during training.
+    """
+    def __init__(self, vocab: Vocabulary,
+                 text_field_embedder: TextFieldEmbedder,
+                 encoder: Seq2SeqEncoder,
+                 softmax: SoftmaxWithNLL,
+                 proj: bool = False,
+                 relu: bool = False,
+                 dropout: float = 0.0,
+                 batch_first: bool=True,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+
+        super().__init__(vocab, regularizer)
+
+        self.encoder = encoder
+        self.batch_first = batch_first
+        self.text_field_embedder = text_field_embedder
+
+        if dropout:
+            self.dropout = torch.nn.Dropout(dropout)
+            dropout_list = [self.dropout]
+        else:
+            self.dropout = None
+            dropout_list = []
+
+        if proj:
+            self.proj = nn.Linear(self.encoder.get_output_dim(), softmax.get_input_dim())
+            proj_list = [self.proj] 
+            if relu:
+                proj_list += [nn.ReLU()]
+            proj_list = proj_list + dropout_list
+        else:
+            self.proj = None
+            proj_list = []
+
+        after_encoder_list = dropout_list + proj_list
+        if len(after_encoder_list) > 0:
+            self.after_encoder = nn.Sequential(*after_encoder_list)
+        else:
+            self.after_encoder = None
+
+        self.softmax = softmax
+        if not self.softmax.adaptive and self.softmax.head.weight.size() == self.text_field_embedder.token_embedder_tokens.weight.size():
+            self.softmax.head.weight = self.text_field_embedder.token_embedder_tokens.weight
+
+        self.softmax_in = softmax.get_input_dim()
+
+        self.metrics = {"ppl": Perplexity()}
+
+        check_dimensions_match(text_field_embedder.get_output_dim(), 
+            encoder.get_input_dim(), 
+            "text field embedder output dim",
+            "encoder input dim")
+        if not proj:
+            check_dimensions_match(encoder.get_output_dim(), 
+                self.softmax_in,
+                "encoder output dim",
+                "softmax input dim")
+
+        initializer(self)
+
+    @overrides
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        return {metric_name: metric.get_metric(reset) for metric_name, metric in self.metrics.items()}
+        
+    @overrides
+    def train(self, mode=True):
+        self.encoder.reset_states()
```

Comment:
Why do you `reset_states` here?

Reply:
I'm not sure whether it's the standard practice for language modeling. But at the start of evaluation or a new training epoch, i would reset the hidden states to all zeros. Since I'm using the default trainer, I choose to add ```reset_states()``` here as a workaround.